tests/cred_mgmt.rs

```diff
@@ -0,0 +1,139 @@
+        ..Default::default()
+    };
+    virt::run_ctap2_with_options(options, |device| {
+        let mut authenticator = Authenticator::new(device).set_pin(b"123456");
+        let metadata = authenticator.credentials_metadata();
+        assert_eq!(metadata.existing, 0);
+        // TODO: check metadata.remaining -- currently not checking the config properly
+
+        for i in 0..10 {
+            let rp = Rp::new(format!("rp{i}"));
+            let user = User::new(b"john.doe");
+            authenticator.make_credential(rp, user).unwrap();
+
+            let metadata = authenticator.credentials_metadata();
+            assert_eq!(metadata.existing, i + 1);
+            // TODO: check metadata.remaining
+        }
+
+        let rps = authenticator.list_rps();
+        assert_eq!(rps.len(), 10);
+        for rp in &rps {
+            let credentials = authenticator.list_credentials(&rp.id);
+            assert_eq!(credentials.len(), 1);
+        }
+
+        let rp = Rp::new("rp11");
+        let user = User::new(b"john.doe");
+        let result = authenticator.make_credential(rp, user);
+        assert_eq!(result, Err(Ctap2Error(0x28)));
+
+        let rps = authenticator.list_rps();
+        assert_eq!(rps.len(), 10);
+        for rp in &rps {
+            let credentials = authenticator.list_credentials(&rp.id);
+            assert_eq!(credentials.len(), 1);
+        }
+    })
+}
+
+#[test]
+fn test_filesystem_full() {
+    let mut options = Options {
+        max_resident_credential_count: Some(10),
+        ..Default::default()
+    };
+    for i in 0..80 {
+        let path = PathBuf::from(format!("/test/{i}").as_str());
+        options.files.push((path, vec![0; 512]));
+    }
+    // TODO: inspect filesystem after run and check remaining blocks
+    virt::run_ctap2_with_options(options, |device| {
+        let mut authenticator = Authenticator::new(device).set_pin(b"123456");
+        let metadata = authenticator.credentials_metadata();
+        assert_eq!(metadata.existing, 0);
+        // This number depends on filesystem layout details and may change if the filesystem
+        // layout or implementation are changed.
+        assert_eq!(metadata.remaining, 5);
+        let n = metadata.remaining;
+
+        let mut i = 0;
+        loop {
+            let rp = Rp::new(format!("rp{i}"));
+            let user = User::new(b"john.doe");
+            let result = authenticator.make_credential(rp, user);
+
+            if result == Err(Ctap2Error(0x28)) {
+                break;
+            }
+            result.unwrap();
+
+            let metadata = authenticator.credentials_metadata();
+            assert_eq!(metadata.existing, i + 1);
+
+            i += 1;
+        }
+
+        // We should be able to create at least 1 but not more than n credentials.
+        assert!(i > 0);
+        assert!(i < n);
+        // Our estimate should not be more than one credential off.
+        assert!(n - i <= 1);
```

Comment:
These feel a bit strict. But given that the test run in a controlled storage and everything always has the same size it might work?

If we ever need to relax this I would relax it a lot more (the remaining estimate should not be off by more than 20% for some size of credential).


It would be interesting to test with very large credentials too (here the estimate will be way higher than the real count).

Reply:
The initial estimate is only 5 with this setup, so this is equivalent to 20 %.  We could also allow `n + 1` but I’d rather have a strict check and relax it if necessary so that we notice if we change the behavior of the estimation.